src/client.ts

```diff
@@ -507,14 +510,23 @@ export class StreamChat<
     this.userID = user.id;
     this.anonymous = false;
 
-    const setTokenPromise = this._setToken(user, userTokenOrProvider);
-    this._setUser(user);
+    const startConnecting = () => {
+      const setTokenPromise = this._setToken(user, userTokenOrProvider);
+      this._setUser(user);
 
-    const wsPromise = this.openConnection();
+      const wsPromise = this.openConnection();
 
-    this.setUserPromise = Promise.all([setTokenPromise, wsPromise]).then(
-      (result) => result[1], // We only return connection promise;
-    );
+      return Promise.all([setTokenPromise, wsPromise]).then(
+        (result) => result[1], // We only return connection promise;
+      );
+    };
+
+    this.setUserPromise = this.pendingDisconnect
+      ? this.pendingDisconnect.then(() => {
+          this.pendingDisconnect = null;
+          return startConnecting();
+        })
+      : startConnecting();
```

Comment:
We could do either this or `this.pendingDisconnect` could be always set to some kind of a Promise (`Promise.resolve()` for example) and that way I wouldn't have to check whether it's `null` or not (also `startConnecting` wouldn't have to exist).
@petyosi @szuperaz